blog/fast-portfolio-sorts/index.qmd

```diff
@@ -0,0 +1,178 @@
+  aggregate(ret_excess ~ portfolio, data = df, FUN = function(x) mean(x, na.rm = TRUE))
+}
+sort_base()
+```
+
+This approach takes about 10 seconds per execution on my machine and is in fact more than 20-times slower than the other approaches! To create a more nuanced picture for the fast and arguably more interesting approaches, I'll drop the `base` approach going forward. 
+
+If we want to perform the same logic using `dplyr`, we can use the following approach. Note that I use `as.data.frame()` for all approaches to ensure that the output format is the same for all approaches - a necessary requirement for a meaningful benchmark (otherwise code would not be equivalent).
+
+```{r}
+sort_dplyr <- function() {
+  crsp_monthly_dplyr |> 
+    mutate(
+      portfolio = assign_portfolio(
+        pick(everything()), "mktcap_lag", n_portfolios = 10),
+      by = "month"
+    ) |> 
+    group_by(portfolio) |> 
+    summarize(ret = mean(ret_excess, na.rm = TRUE)) |> 
+    as.data.frame()
+}
+sort_dplyr()
+```
+
+The equivalent approach in `data.table` looks as follows. Note that I deliberately don't use any pipe or intermediate assignments as to avoid any performance overhead that these might introduce. 
+
+```{r}
+sort_dt <- function() {
+  as.data.frame(crsp_monthly_dt[
+    , `:=`(portfolio = assign_portfolio(.SD, "mktcap_lag", n_portfolios = 10), by = month)
+    ][, .(ret = mean(ret_excess, na.rm = TRUE)), keyby = .(portfolio)])
+}
+sort_dt()
+```
+
+Lastly, I add the `dtplyr` implementation that also takes a `data.table` as input and internally converts `dplyr` code to `data.table` syntax. Note that the final `as.data.frame()` call is used to access the results and ensure that the result format is consistent with the other approaches. 
+
+```{r}
+sort_dtplyr <- function() {
+  crsp_monthly_dtplyr |> 
+    mutate(
+      portfolio = assign_portfolio(
+        pick(everything()), "mktcap_lag", n_portfolios = 10),
+      by = "month"
+    )  |> 
+    group_by(portfolio) |> 
+    summarize(ret = mean(ret_excess, na.rm = TRUE)) |> 
+    as.data.frame()
+}
+sort_dtplyr()
+```
+
+Now that we have verified that all code chunks create the same average excess returns per portfolio, we can proceed to the performance evaluation. 
+
+## Benchmarking results
+
+The `bench` package is a great utility for benchmarking and timing expressions in R. It provides functions that allow you to measure the execution time of  expressions or code chunks. This can be useful for comparing the performance of different approaches or implementations, or for identifying potential bottlenecks in your code. The following code evaluates each approach from above a 100 times and collects the results. 
+
+```{r}
+#| warning: false
+#| eval: false
+set.seed(1234)
+iterations <- 100
+
+results <- bench::mark(
+  sort_dplyr(), sort_dt(), sort_dtplyr(), 
+  iterations = iterations
+)
+```
+
+```{r}
+#| echo: false 
+#| eval: true
+results <- readRDS("data/bench_results.rds") 
```

Comment:
As a side note: What is the right way to handle stuff like this? Seems error prone to me to not include a writeRDS in the part above because even if you sometimes run the script with the benchmarking, you may forget to store bench_results.rds manually.

Reply:
I'll try with cache, but it's also risky to some extent